pennylane_lightning/src/tests/Test_AVXSingleQubitGateHelpers.cpp

```diff
@@ -0,0 +1,344 @@
+        return {"applyExternal", rev_wire, inverse};
+    }
+};
+
+TEMPLATE_TEST_CASE("Test SingleQubitGateHelper template functions",
+                   "[SingleQubitGateHelper]", float, double) {
+    STATIC_REQUIRE(HasInternalWithoutParam<
+                   MockSingleQubitGateWithoutParam<TestType, 4>>::value);
+    STATIC_REQUIRE(!HasInternalWithParam<
+                   MockSingleQubitGateWithoutParam<TestType, 4>>::value);
+    STATIC_REQUIRE(HasExternalWithoutParam<
+                   MockSingleQubitGateWithoutParam<TestType, 4>>::value);
+    STATIC_REQUIRE(!HasExternalWithParam<
+                   MockSingleQubitGateWithoutParam<TestType, 4>>::value);
+
+    STATIC_REQUIRE(!HasInternalWithoutParam<
+                   MockSingleQubitGateWithParam<TestType, 4>>::value);
+    STATIC_REQUIRE(
+        HasInternalWithParam<MockSingleQubitGateWithParam<TestType, 4>>::value);
+    STATIC_REQUIRE(!HasExternalWithoutParam<
+                   MockSingleQubitGateWithParam<TestType, 4>>::value);
+    STATIC_REQUIRE(
+        HasExternalWithParam<MockSingleQubitGateWithParam<TestType, 4>>::value);
+
+    STATIC_REQUIRE(HasInternalWithoutParam<
+                   MockSingleQubitGateSomethingWrong<TestType, 4>>::value);
+    STATIC_REQUIRE(!HasInternalWithParam<
+                   MockSingleQubitGateSomethingWrong<TestType, 4>>::value);
+    STATIC_REQUIRE(!HasExternalWithoutParam<
+                   MockSingleQubitGateSomethingWrong<TestType, 4>>::value);
+    STATIC_REQUIRE(HasExternalWithParam<
+                   MockSingleQubitGateSomethingWrong<TestType, 4>>::value);
+
+    // Test concepts
+    STATIC_REQUIRE(SingleQubitGateWithoutParam<
+                   MockSingleQubitGateWithoutParam<TestType, 4>>);
+    STATIC_REQUIRE(
+        SingleQubitGateWithParam<MockSingleQubitGateWithParam<TestType, 4>>);
+}
+
+TEMPLATE_TEST_CASE("Test SingleQubitGateWithoutParamHelper",
+                   "[SingleQubitGateHelper]", float, double) {
+    auto fallback = [](std::complex<TestType> *arr, size_t num_qubits,
+                       const std::vector<size_t> &wires,
+                       bool inverse) -> std::tuple<std::string, size_t, bool> {
+        static_cast<void>(arr);
+        return {"fallback", num_qubits - wires[0] - 1, inverse};
+    };
+    SECTION("Test SingleQubitGateWithoutParamHelper with packed_size = 4") {
+        constexpr size_t packed_size = 4;
+        std::vector<std::complex<TestType>> arr(
+            16, std::complex<TestType>{0.0, 0.0});
+        SingleQubitGateWithoutParamHelper<
+            MockSingleQubitGateWithoutParam<TestType, packed_size>>
+            func(fallback);
+        // We pack 4 real numbers -> 2 complex numbers -> single qubit.
+        // Thus only rev_wire = 0 calls the internal functions.
+
+        for (bool inverse : {false, true}) {
+            { // num_qubits= 4, wires = {0} -> rev_wires = 3
+                const auto res = func(arr.data(), 4, {0}, inverse);
```

Comment:
Maybe I missed it earlier, but is the signature of the `func` functions defined, so we know the positional data in the tuple?

Reply:
Yeap, `func` returns exactly same type as `applyInternal`, `applyExternal`, and `fallback`.